src/main/java/org/spongepowered/common/text/conversation/SpongeConversation.java

```diff
@@ -0,0 +1,256 @@
+import org.spongepowered.api.text.conversation.EndingHandler;
+import org.spongepowered.api.text.conversation.ExternalChatHandler;
+import org.spongepowered.api.text.conversation.Question;
+import org.spongepowered.api.data.DataContainer;
+import org.spongepowered.api.data.MemoryDataContainer;
+import org.spongepowered.api.event.SpongeEventFactory;
+import org.spongepowered.api.event.cause.Cause;
+import org.spongepowered.api.event.cause.NamedCause;
+import org.spongepowered.api.event.cause.conversation.ConversationEndCause;
+import org.spongepowered.api.event.cause.conversation.ConversationEndType;
+import org.spongepowered.api.event.conversation.ConversationCloseEvent;
+import org.spongepowered.api.plugin.PluginContainer;
+import org.spongepowered.common.SpongeImpl;
+
+import java.lang.ref.WeakReference;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.WeakHashMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+
+public class SpongeConversation implements Conversation {
+
+    private final PluginContainer creator;
+    private final ConversationArchetype archetype;
+
+    private final Map<WeakReference<Conversant>, ExternalChatHandler> handlers = new ConcurrentHashMap<>();
+    private final Map<Conversant, ExternalChatHandler> externalChatHandlers;
+    private final DataContainer context = new MemoryDataContainer();
+    private final Set<EndingHandler> endingHandlers;
+    @Nullable private Question currentQuestion;
+    private boolean catchesOutput;
+    private boolean allowCommands;
+
+    private boolean ended;
+
+    /**
+     * Creates a new conversation.
+     *
+     * @param archetype The archetype for the question
+     * @param conversants All conversants in the conversation
+     * @param creator Who created the conversation
+     */
+    SpongeConversation(ConversationArchetype archetype, Collection<Conversant> conversants, PluginContainer creator) {
+        this.archetype = archetype;
+        this.creator = creator;
+        this.endingHandlers = new HashSet<>(archetype.getEndingHandlers());
+        this.catchesOutput = archetype.catchesOutput();
+        this.allowCommands = archetype.allowsCommands();
+        this.currentQuestion = archetype.getFirstQuestion();
+        this.externalChatHandlers = Collections.synchronizedMap(new WeakHashMap<>());
+
+        conversants.forEach(c -> this.externalChatHandlers.put(c, archetype.getDefaultChatHandler()));
+
+        Task.builder()
+                .async()
+                .interval(5, TimeUnit.MINUTES)
+                .execute(() -> {
+                    if (this.externalChatHandlers.size() == 0) {
+                        SpongeImpl.getLogger().error("A " + creator.getId() + ":" + getId().toLowerCase() + " conversation has ended due to being "
```

Comment:
Add a locale for the call to `toLowerCase()` (usually `Locale.ROOT` is appropriate for these)